spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -0,0 +1,169 @@
+    // in the example below, when id = 5, both actions match but the first one is applied
+    //   WHEN MATCHED AND id > 1 AND id < 10 UPDATE *
+    //   WHEN MATCHED AND id = 5 OR id = 21 DELETE
+
+    val pair = actions.find {
+      case (predicate, _) => predicate.eval(inputRow)
+    }
+
+    // apply the projection to produce an output row, or return null to suppress this row
+    pair match {
+      case Some((_, Some(projection))) =>
+        projection.apply(inputRow)
+      case _ =>
+        null
+    }
+  }
+
+  private def processPartition(
+      params: MergeRowsParams,
+      rowIterator: Iterator[InternalRow]): Iterator[InternalRow] = {
+
+    val joinedAttrs = params.joinedAttributes
+    val isSourceRowPresentPred = createPredicate(params.isSourceRowPresent, joinedAttrs)
+    val isTargetRowPresentPred = createPredicate(params.isTargetRowPresent, joinedAttrs)
+    val matchedPreds = params.matchedConditions.map(createPredicate(_, joinedAttrs))
+    val matchedProjs = params.matchedOutputs.map(_.map(createProjection(_, joinedAttrs)))
+    val notMatchedPreds = params.notMatchedConditions.map(createPredicate(_, joinedAttrs))
+    val notMatchedProjs = params.notMatchedOutputs.map(_.map(createProjection(_, joinedAttrs)))
+    val projectTargetCols = createProjection(params.targetOutput, joinedAttrs)
+    val nonMatchedPairs = notMatchedPreds zip notMatchedProjs
+    val matchedPairs = matchedPreds zip matchedProjs
+    val rowIdAttrs = params.rowIdAttrs
+    val rowIdProj = createProjection(rowIdAttrs, joinedAttrs)
+    val performCardinalityCheck = params.performCardinalityCheck
+    val emitNotMatchedTargetRows = params.emitNotMatchedTargetRows
+
+    // This method is responsible for processing a input row to emit the resultant row with an
+    // additional column that indicates whether the row is going to be included in the final
+    // output of merge or not.
+    // 1. Found a target row for which there is no corresponding source row (join condition not met)
+    //    - Only project the target columns if we need to output unchanged rows
+    // 2. Found a source row for which there is no corresponding target row (join condition not met)
+    //    - Apply the not matched actions (i.e INSERT actions) if non match conditions are met.
+    // 3. Found a source row for which there is a corresponding target row (join condition met)
+    //    - Apply the matched actions (i.e DELETE or UPDATE actions) if match conditions are met.
+    def processRow(inputRow: InternalRow): InternalRow = {
+      if (emitNotMatchedTargetRows && !isSourceRowPresentPred.eval(inputRow)) {
+        projectTargetCols.apply(inputRow)
+      } else if (!isTargetRowPresentPred.eval(inputRow)) {
+        applyProjection(nonMatchedPairs, inputRow)
+      } else {
+        applyProjection(matchedPairs, inputRow)
+      }
+    }
+
+    var lastMatchedRowId: InternalRow = null
+
+    def processRowWithCardinalityCheck(inputRow: InternalRow): InternalRow = {
```

Comment:
I am going to add tests for different join types.

Reply:
Can we test a case where the target table small and the source table is large to attempt to trigger the left broadcast?

Reply:
Will add.